File: PlanningService/src/main/java/org/example/planningservice/commons/result/ServiceError.java
package org.example.planningservice.commons.result;

import lombok.Getter;

import java.util.Map;

@Getter
public class ServiceError {
    public static final String NOT_FOUND = "NotFoundError";
    public static final String UNHANDLED = "InternalError";
    public static final String VALIDATION = "ValidationError";
    public static final String EXISTED = "EntityExistedError";
    public static final String UNAUTHORIZED = "InvalidPermissionError";

    public static final ServiceError NONE = new ServiceError("");

    private final String code;
    private final String description;
    private final Map<String, Object> information;

    public ServiceError(String code) {
        this(code, null, null);
    }

    public ServiceError(String code, String description) {
        this(code, description, null);
    }

    public ServiceError(String code, String description, Map<String, Object> information) {
        this.code = code;
        this.description = description;
        this.information = information;
    }

    public static ServiceError notFoundError(String description) {
        return new ServiceError(NOT_FOUND, description);
    }

    public static ServiceError unhandledException(String description) {
        return new ServiceError(UNHANDLED, description);
    }

    public static ServiceError validationFailed(String description) {
        return new ServiceError(VALIDATION, description);
    }

    public String getCode() {
        return code;
    }

    public Map<String, Object> getInformation() {
        return information;
    }

    public String getDescription() {
        return description;
    }

    public static ServiceError existedError(String description) {
        return new ServiceError(EXISTED, description);
    }

    public static ServiceError unauthorizedError(String description) {
        return new ServiceError(UNAUTHORIZED, description);
    }
}

File: AuthService/Application/Commons/ServiceError.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Application.Commons
{
    public sealed record ServiceError(string Code, string? Description = null, Dictionary<string, object>? Information = null)
    {
        public const string NotFound = "NotFoundError";
        public const string Unhandled = "InternalError";
        public const string Validation = "ValidationError";
        public const string Existed = "EntityExistedError";
        public const string Unauthorized = "InvalidPermissionError";    

        public static readonly ServiceError None = new(string.Empty);

        public static ServiceError NotFoundError(string description) => new(NotFound, description);
        public static ServiceError UnhandledException(string description) => new(Unhandled, description);
        public static ServiceError ValidationFailed(string description) => new(Validation, description);
        public static ServiceError ExistedError(string description) => new(Existed, description);
        public static ServiceError UnauthorizedError(string description) => new(Unauthorized, description);
    }
}
